text/1011-object-level-security.rst

```diff
@@ -0,0 +1,320 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2022-04-18
+    Authors: Elvis Pranskevichus <elvis@edgedb.com>
+    RFC PR: `edgedb/rfcs#0054 <https://github.com/edgedb/rfcs/pull/54>`_
+
+===============================
+RFC 1011: Object-Level Security
+===============================
+
+This RFC proposes adding a mechanism to declare per-type access control
+rules in schemas to automatically enforce data access and modification on
+per-object level.  This is analogous to row-level security (RLS) in SQL
+systems.
+
+
+Motivation
+==========
+
+Object-level security is a powerful feature which allows universal leak-proof
+enforcement of data access policies across all uses of a database.  This is
+important for security and compliance, but would also be tremendously useful
+for backend-less applications that interact with EdgeDB via HTTP (either via
+GraphQL or EdgeQL-over-HTTP).  A secondary motivation is that this will allow
+implementation of `temporal databases <temporal>`_.
+
+
+Requirements
+============
+
+Although there is no direct dependency, this RFC requires
+`RFC 1010 <1001-global-vars.rst>`_ (globals) to be implemented to be practical.
+See `Interaction with globals`_ below for details.
+
+
+Specification
+=============
+
+This RFC proposes to add a new schema item type -- *Access Policy* -- that is
+defined in the context of an object type and specifies the rules of rewriting
+queries that refer to the enclosing object type.
+
+CREATE ACCESS POLICY
+--------------------
+
+Define a new access control policy for a given object type.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    {CREATE|ALTER} OBJECT TYPE <type-name> "{"
+        CREATE ACCESS POLICY <name>
+            [ WHEN <condition> ]
+            { PERMIT | RESTRICT }
+            { ALL | READ | WRITE | DELETE } [ ... ]
+            USING (<expr>)
+    "}"
+
+An access policy grants the permission to select, insert or delete objects
+that match the relevant policy expression.  The ``READ`` and ``DELETE``
+policies specify a *visibility* check for existing objects: when ``<expr>>``
+returns a false result for an object, such object will be silently skipped
+in any ``SELECT``, ``UPDATE`` or ``DELETE`` expression that scans it.
+A ``WRITE`` policy specifies a *validity* check for new or updated objects and
+affects ``INSERT`` and ``UPDATE`` expressions: if the proposed object violates
+``<expr>``, an error is raised immediately and the query is aborted.
+
+The optional ``<condition>`` expression is evaluated for every object
+affected by the statement and the policy is applied only if the expression
+evaluates to *true*.  It is essentially equivalent to joining ``<condition>``
+with ``<expr>`` with an ``AND`` operator.  The reason for a standalone clause
```

Comment:
Is `<condition>` negated before `AND`ing with `<expr>` for `RESTRICT` queries, then? To match the intuition that the policy is only applied then?

Reply:
Yeah, the wording in this paragraph is wrong, the equivalent of `when <condition>` is `not <condition> or <policy-expr>`

Reply:
It's still `AND` for `PERMIT` though, right?

Reply:
No.  `PERMIT` is combined with `OR`

Reply:
Oh, I misunderstood your question.  Yes, your understanding is correct.  `permit` does *not* "fail open".  In other words, if `<condition>` is false, no objects will be added to the "permitted" set.

Reply:
Looking at the examples again, the use of `when` for the `age_appropriate` access policy seems off. That rule probably ought to be `restrict to`? (Which then would require a catch all `permit` rule?)